### DISCUSSION TITLE
DynamicForm module

### DISCUSSION THREAD

**Discussion by leigh-pointer** [oqtane/oqtane.framework#2051] [Show and tell]
My new module Dynamic Form is available for download.

This is a preview version but fully functional.  Documentation and Forms library will also be available in the very near future. 

"StudioElf.DynamicForm - Is a Forms Designer lets you easily add controls to a form, and arrange them.  Once your form is designed you are able to capture and display data in a presenter"

Please give feedback.

![image](https://user-images.githubusercontent.com/9447612/157300391-c704a5cf-45d7-4eb5-993a-9ccdfd887437.png)

Example Forms (applets)
Task List
![image](https://user-images.githubusercontent.com/9447612/157369985-f0296545-338a-4793-96e8-541675e30588.png)

![image](https://user-images.githubusercontent.com/9447612/157370049-cf8105a1-474e-4a40-8afc-2e5783c29a97.png)

![image](https://user-images.githubusercontent.com/9447612/157370105-9196a7f3-cd7d-4563-8ead-a08369495de3.png)

![image](https://user-images.githubusercontent.com/9447612/157370150-64f457da-a669-4360-96d9-d38d67601f96.png)

![image](https://user-images.githubusercontent.com/9447612/157370226-f63faf4e-23e2-4a3f-8cfd-be98503e5f1e.png)

Blog
![image](https://user-images.githubusercontent.com/9447612/157370318-dae7de1a-a344-40a0-b103-b86f2eebe836.png)

Inline add and edit (no modal)
![image](https://user-images.githubusercontent.com/9447612/157370501-47064e07-1856-4cf9-a485-f21edb46da36.png)

Address Book
![image](https://user-images.githubusercontent.com/9447612/157370578-e5cbdcf0-0c4b-4665-9d65-7bcaa7eb74b0.png)

The possibilities are endless.

All Forms can be shared through file import or URL with or without data.
Data can be consumed thru a restAPI (readonly)

**Comment by ghhv** [comment]
Looks great - started testing.. 

FYI - typo on "syncoronize"

![image](https://user-images.githubusercontent.com/13160762/157794863-6b3bcbbe-d26b-4aca-af4d-9d738836a545.png)

FYI 2: And I think correct English spelling is generally synchronise/ize with the 'h'.

**Reply by ghhv** [reply]
Hi Leigh, 
This is incredible. The functionality is amazing for a form creator. 
I'm a bit stuck though.. I've created a form but can't seem to display it on the page as populatable content. I can switch the Designer to "User Compose" mode and see a table, I can click "Add New" and fill out the form but the data never appears. 
So how does one turn off the designer buttons and display user compose mode only?
e.g.
![image](https://user-images.githubusercontent.com/13160762/157800256-4ee6f573-bb73-4cd1-ab29-6af0e822613a.png)
![image](https://user-images.githubusercontent.com/13160762/157800298-63539bbc-a9fa-4ec7-9091-8d22fc8dccca.png)

Or maybe I have this wrong - is this something you can use to create forms to use in your own modules?  If so, how does that step work? 
Is the data accessible from the database between modules or other projects or only via the read only API you mentioned?
Cheers, Harry

**Reply by leigh-pointer** [reply]
@ghhv 
> This is incredible. The functionality is amazing for a form creator.
Thanks.
The module requires 3 major steps, Create your form, add data and design the presenter.
The presenter designer is raw html 
![image](https://user-images.githubusercontent.com/9447612/157829004-7fde5209-868a-4783-89d4-d5d5796e5fb2.png)

![image](https://user-images.githubusercontent.com/9447612/157828303-905ed59c-337e-4228-85e2-c95c1b1fbf32.png)


Here is an AddressBook module with data.  
![image](https://user-images.githubusercontent.com/9447612/157828842-41006aaf-58ec-4b04-a301-ce24daa6a43c.png)


To import this you will need to update the module settings and make sure the Import Data is checked.
![image](https://user-images.githubusercontent.com/9447612/157829374-d1449d35-67b9-46cc-90a5-12ab5069b268.png)

Unzip the AddressBook.zip and import the json file 
![image](https://user-images.githubusercontent.com/9447612/157829744-0565e261-a7c1-4193-a148-98c3f813a160.png)

[Address Book.zip](https://github.com/oqtane/oqtane.framework/files/8230261/Address.Book.zip)

**Reply by leigh-pointer** [reply]
> So how does one turn off the designer buttons and display user compose mode only?
On the module permissions tab select the role that can compose 
![image](https://user-images.githubusercontent.com/9447612/157832795-2b5653af-4e74-4c6a-a7c8-438afb0a0c2d.png)
this must be save and settings reopened to allow the next step.
Open the module settings and select what and how the Role can use the data.
![image](https://user-images.githubusercontent.com/9447612/157833303-4a507edc-ea22-4e13-8648-b661d7b3b248.png)
When a user of that role is logged in the Toolbar is not available unless they have the designer permission.

**Reply by leigh-pointer** [reply]
> Is the data accessible from the database between modules or other projects or only via the read only API you mentioned?
Data is only accessible through the readonly API and the link is shown on the module settings tab [Endpoint].

**Reply by ghhv** [reply]
Whoa.. that was a deeper answer than I expected.. Thanks. Shall review again tomorrow!

**Reply by leigh-pointer** [reply]
Let me know how you get on

**Reply by drosenba** [reply]
Sorry, I'm new to Oqtane, ...
RE: "...update the module settings..."
Where is that?
RE: "...On the module permissions tab..."
Where is that?

I'm not understanding the 'workflow' to create multiple/different forms, and 'publish' them to different pages (for end users).  
Also, as a form designer user, do I only have access to the one/last form I Import, or is there some way to 'load' a prior form?

**Reply by leigh-pointer** [reply]
@drosenba the module settings and permissions are found using the dropdown menu found usually athe the top right of the module title! You will need to be logged in as admin or host.
The module currently only allows for one form to be generated. Multiple forms and workflow is planned for a future release.
There is nothing to stop you adding another instance of the module and creating a new form. I have built many forms on one page in stress testing.

let me know how you get on and happy to answer any questions.

**Reply by drosenba** [reply]
Thanks Leigh.  OK, things are a little clearer now.  I'm very impressed by the project...

What does permission EditReadOnly mean?


RE: "User Compose":  Maybe it's just a language thing, but the term "Compose" is a little confusing.   Maybe in CRM content creation it's appropriate, but not for users entering data by form filling.  Maybe "Enter Data" is clearer (IMHO :)


FYI, in the "User Compose" section, there are no hover-hints on any of the buttons.  Also, if users don't have the designer permission, they can click the 'Presenter View' button, but when in that mode, I don't see a way to get back to User Compose mode (except to refresh the page).


RE: "adding another instance of the module and creating a new form"
OK, I get it now.  With the data being tied to an _instance_ of the module, then it does not seem that I'd be able to put the same data on another page.  Also, in looking at how the data is stored in database (StudioElf.DynamicForm.Models.DataStore), I might be concerned about performance because the entire set of 'records' is being stored in JSON in a single field.  (I suspect you've considered adding additional features to "Export data to CSV", and "Export data to a DB table" ;)


Will you be releasing the source code?  Open source?  What license type?

**Reply by leigh-pointer** [reply]
@drosenba sorry for late reply im currently taking some R&R 
thanks for your breakdown and will action when sometime next week and deploy a new build.

**EditReadOnly**
You can design your form and you are able to make fields ReadOnly these fields can only be edited by users with the EditReadOnly permission. 
I have done some stress testing on the storage and im very happy with performance. 
Data can be exported by using the RestAPI (readonly) and this will be extended to handle csv.
This will be a commercial module with a restricted free version.

**Comment by vnetonline** [comment]
@leigh-pointer your dynamic forms module is brilliant … are you planning to sell it on market place ?

**Reply by leigh-pointer** [reply]
@vnetonline Thank you, yes the module is planned for the Market Place.  One more Beta Release then it goes Release Candidate.  Have you seen some of the Form / Web Parts on https://www.studio-elf.net/dynamicform/forms

**Reply by vnetonline** [reply]
> @vnetonline Thank you, yes the module is planned for the Market Place.  One more Beta Release then it goes Release Candidate.  Have you seen some of the Form / Web Parts on https://www.studio-elf.net/dynamicform/forms 

Yes I have you could make your own market place and get users of your module to sell their templates … like I said brilliant

**Comment by leigh-pointer** [comment]
@vnetonline 

> Yes I have you could make your own market place and get users of your module to sell their templates

definatly something to consider ;)
